File: internal/docker/stats.go
package docker

import (
	"context"
	"fmt"
	"os/exec"
	"strings"
)

// Stats returns container resource usage for this app's containers.
// It runs docker stats --no-stream filtered to containers with the jib-<app> prefix.
func (c *Compose) Stats(ctx context.Context) (string, error) {
	// Get container IDs matching the project prefix
	psCmd := exec.CommandContext(ctx, "docker", "ps", "-q", "--filter", fmt.Sprintf("label=com.docker.compose.project=%s", c.ProjectName()))
	psOut, err := psCmd.CombinedOutput()
	if err != nil {
		return "", fmt.Errorf("listing containers: %w: %s", err, string(psOut))
	}

	ids := strings.Fields(strings.TrimSpace(string(psOut)))
	if len(ids) == 0 {
		return fmt.Sprintf("No running containers for %s", c.ProjectName()), nil
	}

	args := []string{"stats", "--no-stream", "--format", "table {{.Name}}\t{{.CPUPerc}}\t{{.MemUsage}}\t{{.NetIO}}"}
	args = append(args, ids...)

	cmd := exec.CommandContext(ctx, "docker", args...)
	out, err := cmd.CombinedOutput()
	if err != nil {
		return "", fmt.Errorf("getting stats: %w: %s", err, string(out))
	}

	return strings.TrimSpace(string(out)), nil
}

File: libs/docker/stats.ts
import { JibError } from '@jib/core'
import type { Compose } from './compose.ts'
import { type DockerExec, realExec } from './exec.ts'

/**
 * Returns a human-readable table of container resource usage for this app's
 * containers. Parity with the Go helper: filters by
 * `com.docker.compose.project=<projectName>` then invokes `docker stats`
 * against the matching IDs.
 */
export async function composeStats(compose: Compose, exec: DockerExec = realExec): Promise<string> {
  const project = compose.projectName()
  const ps = await exec(
    ['docker', 'ps', '-q', '--filter', `label=com.docker.compose.project=${project}`],
    { cwd: compose.cfg.dir, capture: true },
  )
  if (ps.exitCode !== 0) {
    throw new JibError('docker.stats', `listing containers failed: ${ps.stderr}`)
  }
  const ids = ps.stdout.split(/\s+/).filter(Boolean)
  if (ids.length === 0) return `No running containers for ${project}`

  const args = [
    'docker',
    'stats',
    '--no-stream',
    '--format',
    'table {{.Name}}\t{{.CPUPerc}}\t{{.MemUsage}}\t{{.NetIO}}',
    ...ids,
  ]
  const res = await exec(args, { cwd: compose.cfg.dir, capture: true })
  if (res.exitCode !== 0) {
    throw new JibError('docker.stats', `stats failed: ${res.stderr}`)
  }
  return res.stdout.trim()
}
